File: sdk/python/src/covenant/interceptor.py
from __future__ import annotations

import contextvars
import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any

from covenant.errors import CovenantViolationError

if TYPE_CHECKING:
    from covenant.models import CovenantSpec

# Per-invocation interceptor — set by ContractEnforcer before awaiting the agent fn.
# Each asyncio task gets its own value; concurrent invocations are fully isolated.
_active_interceptor: contextvars.ContextVar[ToolCallInterceptor | None] = (
    contextvars.ContextVar("_active_interceptor", default=None)
)

# Best-effort price table (input $/1k tokens, output $/1k tokens)
_PRICE_TABLE: dict[str, tuple[float, float]] = {
    "gpt-4o": (0.005, 0.015),
    "gpt-4o-mini": (0.00015, 0.0006),
    "gpt-4-turbo": (0.01, 0.03),
    "gpt-3.5-turbo": (0.0005, 0.0015),
    "claude-opus-4-6": (0.015, 0.075),
    "claude-sonnet-4-6": (0.003, 0.015),
    "claude-haiku-4-5-20251001": (0.00025, 0.00125),
}


@dataclass
class ToolCall:
    """A single tool call observed during an agent invocation.

    Args:
        name: Tool function name.
        args: Arguments passed to the tool.
        result: Return value from the tool (None if not yet resolved).
        cost_usd: LLM cost attributed to this call.
        timestamp: UTC time the call was observed.
    """

    name: str
    args: dict
    result: Any
    cost_usd: float
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


class ToolCallInterceptor:
    """Stateful per-invocation enforcer installed via ContextVar.

    Tracks all tool calls and cumulative cost for a single @contract invocation.
    Raises CovenantViolationError immediately on any constraint breach.
    """

    def __init__(self, spec: CovenantSpec) -> None:
        self._spec = spec
        self.tool_calls: list[ToolCall] = []
        self.total_cost_usd: float = 0.0
        self.model_used: str | None = None

    def record_completion(self, response: Any, model: str) -> None:
        """Called after each OpenAI completion response.

        Extracts usage, computes cost, checks model declaration, checks tool calls
        in the response, and enforces budget + call limits.

        Args:
            response: OpenAI ChatCompletion response object.
            model: Model identifier string from the request kwargs.

        Raises:
            CovenantViolationError: On BUDGET_EXCEEDED, UNDECLARED_MODEL,
                UNDECLARED_TOOL, DENIED_TOOL, or CALL_LIMIT_EXCEEDED.
        """
        self.model_used = model

        # Accumulate cost from usage
        usage = getattr(response, "usage", None)
        if usage:
            in_tok = getattr(usage, "prompt_tokens", 0) or 0
            out_tok = getattr(usage, "completion_tokens", 0) or 0
            in_price, out_price = _PRICE_TABLE.get(model, (0.0, 0.0))
            self.total_cost_usd += (in_tok / 1000 * in_price) + (out_tok / 1000 * out_price)

        # Budget check
        budget = self._spec.constraints.budget
        if budget and self.total_cost_usd > budget.max_cost_usd:
            raise CovenantViolationError(
                code="BUDGET_EXCEEDED",
                detail={
                    "max_cost_usd": budget.max_cost_usd,
                    "actual_cost_usd": round(self.total_cost_usd, 6),
                },
            )

        # Model check (capabilities.models is a list of ModelAllowSpec)
        declared_models = self._spec.capabilities.models
        if declared_models is not None:
            allowed_model_ids = {m.model for m in declared_models}
            if model not in allowed_model_ids:
                raise CovenantViolationError(
                    code="UNDECLARED_MODEL",
                    detail={"model": model, "declared": sorted(allowed_model_ids)},
                )

        # Extract tool calls from response choices
        choices = getattr(response, "choices", None) or []
        if choices:
            msg = getattr(choices[0], "message", None)
            response_tool_calls = getattr(msg, "tool_calls", None) or []
            for tc in response_tool_calls:
                fn = getattr(tc, "function", None)
                if not fn:
                    continue
                tool_name = fn.name
                args = json.loads(fn.arguments) if fn.arguments else {}
                self._check_tool(tool_name)
                self.tool_calls.append(
                    ToolCall(name=tool_name, args=args, result=None, cost_usd=0.0)
                )

        # Call limit check
        scope = self._spec.constraints.scope
        if scope and scope.max_calls_per_invocation is not None:
            if len(self.tool_calls) > scope.max_calls_per_invocation:
                raise CovenantViolationError(
                    code="CALL_LIMIT_EXCEEDED",
                    detail={
                        "max_calls": scope.max_calls_per_invocation,
                        "actual_calls": len(self.tool_calls),
                    },
                )

    def _check_tool(self, name: str) -> None:
        """Verify tool name against capabilities and deny list.

        Raises:
            CovenantViolationError: DENIED_TOOL or UNDECLARED_TOOL.
        """
        allowed = set(self._spec.capabilities.tools)
        denied = set(self._spec.constraints.deny_tools or [])

        if name in denied:
            raise CovenantViolationError(
                code="DENIED_TOOL",
                detail={"tool": name},
            )
        if name not in allowed:
            raise CovenantViolationError(
                code="UNDECLARED_TOOL",
                detail={"tool": name, "declared": sorted(allowed)},
            )


# ---------------------------------------------------------------------------
# OpenAI SDK class-level patch — installed once at module import time.
# The patch is a no-op when no interceptor is active in the current context.
# ---------------------------------------------------------------------------

def _install_patches() -> None:
    """Monkey-patch OpenAI Completions at class level. Idempotent."""
    try:
        import openai.resources.chat.completions.completions as _mod
    except ImportError:
        # openai not installed — skip patching (SDK still usable without LLM interception)
        return

    Completions = _mod.Completions
    AsyncCompletions = _mod.AsyncCompletions

    if getattr(Completions, "_covenant_patched", False):
        return

    _orig_create = Completions.create
    _orig_async_create = AsyncCompletions.create

    def _patched_create(self_inner, *args, **kwargs):
        response = _orig_create(self_inner, *args, **kwargs)
        interceptor = _active_interceptor.get()
        if interceptor is not None:
            interceptor.record_completion(response, kwargs.get("model", ""))
        return response

    async def _patched_async_create(self_inner, *args, **kwargs):
        response = await _orig_async_create(self_inner, *args, **kwargs)
        interceptor = _active_interceptor.get()
        if interceptor is not None:
            interceptor.record_completion(response, kwargs.get("model", ""))
        return response

    Completions.create = _patched_create
    AsyncCompletions.create = _patched_async_create
    Completions._covenant_patched = True
    AsyncCompletions._covenant_patched = True


_install_patches()

File: sdk/typescript/src/interceptor.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { CovenantViolationError } from "./errors.js";
import type { CovenantSpec } from "./spec.js";

// ── Price table (input $/1k tokens, output $/1k tokens) ──────────────────────

const PRICE_TABLE: Record<string, [number, number]> = {
  "gpt-4o": [0.005, 0.015],
  "gpt-4o-mini": [0.00015, 0.0006],
  "gpt-4-turbo": [0.01, 0.03],
  "gpt-3.5-turbo": [0.0005, 0.0015],
  "claude-opus-4-6": [0.015, 0.075],
  "claude-sonnet-4-6": [0.003, 0.015],
  "claude-haiku-4-5-20251001": [0.00025, 0.00125],
};

// ── ToolCall record ───────────────────────────────────────────────────────────

export interface ToolCall {
  name: string;
  args: Record<string, unknown>;
  result: unknown;
  costUsd: number;
  timestamp: Date;
}

// ── Interceptor ───────────────────────────────────────────────────────────────

export class ToolCallInterceptor {
  readonly toolCalls: ToolCall[] = [];
  totalCostUsd = 0;
  modelUsed: string | null = null;

  constructor(private readonly spec: CovenantSpec) {}

  /**
   * Called after each OpenAI completion response.
   * Accumulates cost, checks model declaration, extracts + validates tool calls,
   * enforces budget and call limits.
   */
  recordCompletion(response: unknown, model: string): void {
    this.modelUsed = model;

    // Accumulate cost
    const usage = (response as Record<string, unknown>)["usage"] as
      | Record<string, number>
      | undefined;
    if (usage) {
      const inTok = usage["prompt_tokens"] ?? 0;
      const outTok = usage["completion_tokens"] ?? 0;
      const [inPrice, outPrice] = PRICE_TABLE[model] ?? [0, 0];
      this.totalCostUsd += (inTok / 1000) * inPrice + (outTok / 1000) * outPrice;
    }

    // Budget check
    const budget = this.spec.constraints.budget;
    if (budget && this.totalCostUsd > budget.max_cost_usd) {
      throw new CovenantViolationError("BUDGET_EXCEEDED", {
        max_cost_usd: budget.max_cost_usd,
        actual_cost_usd: Math.round(this.totalCostUsd * 1e6) / 1e6,
      });
    }

    // Model check
    const declaredModels = this.spec.capabilities.models;
    if (declaredModels !== undefined) {
      const allowed = new Set(declaredModels.map((m) => m.model));
      if (!allowed.has(model)) {
        throw new CovenantViolationError("UNDECLARED_MODEL", {
          model,
          declared: [...allowed].sort(),
        });
      }
    }

    // Extract tool calls from response
    const choices = (response as Record<string, unknown>)["choices"] as
      | unknown[]
      | undefined;
    if (choices && choices.length > 0) {
      const msg = (choices[0] as Record<string, unknown>)["message"] as
        | Record<string, unknown>
        | undefined;
      const tcs = (msg?.["tool_calls"] as unknown[] | undefined) ?? [];
      for (const tc of tcs) {
        const fn = (tc as Record<string, unknown>)["function"] as
          | Record<string, unknown>
          | undefined;
        if (!fn) continue;
        const toolName = fn["name"] as string;
        let args: Record<string, unknown> = {};
        try {
          args = JSON.parse((fn["arguments"] as string) ?? "{}") as Record<
            string,
            unknown
          >;
        } catch {
          // malformed args — keep empty
        }
        this.checkTool(toolName);
        this.toolCalls.push({
          name: toolName,
          args,
          result: null,
          costUsd: 0,
          timestamp: new Date(),
        });
      }
    }

    // Call limit check
    const maxCalls = this.spec.constraints.scope?.max_calls_per_invocation;
    if (maxCalls !== undefined && this.toolCalls.length > maxCalls) {
      throw new CovenantViolationError("CALL_LIMIT_EXCEEDED", {
        max_calls: maxCalls,
        actual_calls: this.toolCalls.length,
      });
    }
  }

  checkTool(name: string): void {
    const allowed = new Set(this.spec.capabilities.tools);
    const denied = new Set(this.spec.constraints.deny_tools ?? []);

    if (denied.has(name)) {
      throw new CovenantViolationError("DENIED_TOOL", { tool: name });
    }
    if (!allowed.has(name)) {
      throw new CovenantViolationError("UNDECLARED_TOOL", {
        tool: name,
        declared: [...allowed].sort(),
      });
    }
  }
}

// ── AsyncLocalStorage context ─────────────────────────────────────────────────

/**
 * Per-invocation storage. Each async call chain (one @contract invocation)
 * gets its own interceptor instance — concurrent calls are fully isolated.
 */
export const interceptorStorage = new AsyncLocalStorage<ToolCallInterceptor>();

// ── OpenAI SDK class-level patch ──────────────────────────────────────────────

let _patched = false;

export function installPatches(): void {
  if (_patched) return;
  try {
    // Dynamic require to avoid hard dep — if openai isn't installed, skip
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const openaiModule = require("openai/resources/chat/completions") as {
      Completions: { prototype: { create: (...args: unknown[]) => unknown } };
    };
    const Completions = openaiModule.Completions;
    const origCreate = Completions.prototype.create.bind(
      Completions.prototype
    ) as (...args: unknown[]) => unknown;

    Completions.prototype.create = function (
      this: unknown,
      ...args: unknown[]
    ): unknown {
      const interceptor = interceptorStorage.getStore();
      const result = origCreate.apply(this, args);

      // Handle both sync and async results
      if (result instanceof Promise) {
        return result.then((response) => {
          interceptor?.recordCompletion(
            response,
            (args[0] as Record<string, unknown>)?.["model"] as string ?? ""
          );
          return response;
        });
      }
      interceptor?.recordCompletion(
        result,
        (args[0] as Record<string, unknown>)?.["model"] as string ?? ""
      );
      return result;
    };

    _patched = true;
  } catch {
    // openai not installed — interception is no-op
  }
}
